sdk/core/core/src/az_span.c

```diff
@@ -131,95 +175,128 @@ AZ_NODISCARD az_result az_span_to_uint32(az_span span, uint32_t* out_number)
 
 AZ_NODISCARD az_result az_span_copy(az_span destination, az_span source, az_span* out_span)
 {
-  AZ_PRECONDITION_VALID_SPAN(destination, 0, true);
-  AZ_PRECONDITION_VALID_SPAN(source, 0, true);
+  AZ_PRECONDITION_NOT_NULL(out_span);
+
   int32_t src_len = az_span_length(source);
+  int32_t dest_capacity = az_span_capacity(destination);
 
-  if (az_span_capacity(destination) < src_len)
+  if (dest_capacity < src_len)
   {
     return AZ_ERROR_INSUFFICIENT_SPAN_CAPACITY;
-  };
+  }
 
   uint8_t* ptr = az_span_ptr(destination);
+  if (ptr == NULL)
+  {
+    *out_span = AZ_SPAN_NULL;
+  }
+  else
+  {
+    //TODO: What should be the behavior if source.ptr is NULL?
+    memmove((void*)ptr, (void const*)az_span_ptr(source), (size_t)src_len);
 
-  memmove((void*)ptr, (void const*)az_span_ptr(source), (size_t)src_len);
-
-  *out_span = az_span_init(ptr, src_len, az_span_capacity(destination));
-
+    *out_span = _az_span_init_unchecked(ptr, src_len, dest_capacity);
+  }
   return AZ_OK;
 }
 
-AZ_NODISCARD AZ_INLINE bool should_encode(uint8_t c)
-{
-  switch (c)
-  {
-    case '-':
-    case '_':
-    case '.':
-    case '~':
-      return false;
-    default:
-      return !(('0' <= c && c <= '9') || ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z'));
-  }
-}
+static const uint8_t _is_ascii_byte_url[256] = {
```

Comment:
Wondering if we really need this?  Could you elaborate on the benefit of this versus the previous implementation?

Reply:
Yes, will detail the benefits/comments once it's final. :)